src/analysis/inner_product_space/rayleigh.lean

```diff
@@ -0,0 +1,278 @@
+/-
+Copyright (c) 2021 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth, Frédéric Dupuis
+-/
+import analysis.inner_product_space.calculus
+import analysis.inner_product_space.dual
+import analysis.calculus.lagrange_multipliers
+import linear_algebra.eigenspace
+
+/-!
+# The Rayleigh quotient
+
+The Rayleigh quotient of a self-adjoint operator `T` on an inner product space `E` is the function
+`λ x, ⟪T x, x⟫ / ∥x∥ ^ 2`.
+
+The main results of this file are `is_self_adjoint.has_eigenvector_of_is_max_on` and
+`is_self_adjoint.has_eigenvector_of_is_min_on`, which state that if `E` is complete, and if the
+Rayleigh quotient attains its global maximum/minimum over some sphere at the point `x₀`, then `x₀`
+is an eigenvector of `T`, and the `supr`/`infi` of `λ x, ⟪T x, x⟫ / ∥x∥ ^ 2` is the corresponding
+eigenvalue.
+
+The corollaries `is_self_adjoint.has_eigenvalue_supr_of_finite_dimensional` and
+`is_self_adjoint.has_eigenvalue_supr_of_finite_dimensional` state that if `E` is finite-dimensional
+and nontrivial, then `T` has some (nonzero) eigenvectors with eigenvalue the `supr`/`infi` of
+`λ x, ⟪T x, x⟫ / ∥x∥ ^ 2`.
+
+## TODO
+
+A slightly more elaborate corollary is that if `E` is complete and `T` is a compact operator, then
+`T` has some (nonzero) eigenvector with eigenvalue either `⨆ x, ⟪T x, x⟫ / ∥x∥ ^ 2` or
+`⨅ x, ⟪T x, x⟫ / ∥x∥ ^ 2` (not necessarily both).
+
+-/
+
+variables {𝕜 : Type*} [is_R_or_C 𝕜]
+variables {E : Type*} [inner_product_space 𝕜 E]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 _ _ x y
+open_locale nnreal
+
+open module.End metric
+
+namespace continuous_linear_map
+variables (T : E →L[𝕜] E)
+
```

Comment:
Do you think it would ease readability to introduce either:
```lean
private noncomputable abbreviation rayleigh_quotient : E → ℝ :=
λ x : E, T.re_apply_inner_self x / ∥(x:E)∥ ^ 2
```
or even just 
```lean
local notation `T.rayleigh_quotient` := λ x : E, T.re_apply_inner_self x / ∥(x:E)∥ ^ 2
```
for use below?

Reply:
Thanks, I've done the latter.